[Tabs] Adds justified and scrollable layout to MDCTabBarView

leonmz opened the pull request:
Adds justified and scrollable layout for MDCTabBarView and corresponding tests.

## Example View Controller

|Before|After (Portrait)|After (Landscape)
|---|---|---|
|![60151483-ea246900-97aa-11e9-9c1e-688f1d241512](https://user-images.githubusercontent.com/1753199/60179832-920e5680-97ec-11e9-8131-ebcf17f853de.png)|![Simulator Screen Shot - iPhone 7 - 2019-06-26 at 08 27 47](https://user-images.githubusercontent.com/1753199/60179840-98043780-97ec-11e9-9d3a-cbcfd48cb202.png)|![Simulator Screen Shot - iPhone 7 - 2019-06-26 at 08 29 09](https://user-images.githubusercontent.com/1753199/60179813-815de080-97ec-11e9-9d7b-14e19fae4df1.png)|


Part of Issue #7645

Commit:
Adds stack view

Commit:
Adds stack view

material-automation commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -85,32 +73,71 @@ - (void)setSelectedItem:(UITabBarItem *)selectedItem {
     NSString *itemTitle = selectedItem.title;
     NSString *exceptionMessage =
-        [NSString stringWithFormat:@"%@ is not a member of the tab bar's `items`.", itemTitle];
+    [NSString stringWithFormat:@"%@ is not a member of the tab bar's `items`.", itemTitle];
```
clang-format suggested the following change:


```
-    [NSString stringWithFormat:@"%@ is not a member of the tab bar's `items`.", itemTitle];
+        [NSString stringWithFormat:@"%@ is not a member of the tab bar's `items`.", itemTitle];
```

<!-- identifier: clang-format -->
<!-- File: components/Tabs/src/TabBarView/MDCTabBarView.m -->

material-automation commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -85,32 +73,71 @@ - (void)setSelectedItem:(UITabBarItem *)selectedItem {
+                                            [_stackView.widthAnchor constraintGreaterThanOrEqualToAnchor:self.widthAnchor],
+                                            [_stackView.topAnchor constraintEqualToAnchor:self.topAnchor],
+                                            [_stackView.bottomAnchor constraintEqualToAnchor:self.bottomAnchor],
+                                            ]];
```
clang-format suggested the following change:


```
-                                            [self.leadingAnchor constraintEqualToAnchor:self.leadingAnchor],
-                                            [self.trailingAnchor constraintEqualToAnchor:self.trailingAnchor],
-                                            [self.topAnchor constraintEqualToAnchor:self.topAnchor],
-                                            [self.bottomAnchor constraintEqualToAnchor:self.bottomAnchor],
-
-                                            [_stackView.leadingAnchor constraintEqualToAnchor:self.leadingAnchor],
-                                            [_stackView.trailingAnchor constraintEqualToAnchor:self.trailingAnchor],
-                                            [_stackView.widthAnchor constraintGreaterThanOrEqualToAnchor:self.widthAnchor],
-                                            [_stackView.topAnchor constraintEqualToAnchor:self.topAnchor],
-                                            [_stackView.bottomAnchor constraintEqualToAnchor:self.bottomAnchor],
-                                            ]];
+    [self.leadingAnchor constraintEqualToAnchor:self.leadingAnchor],
+    [self.trailingAnchor constraintEqualToAnchor:self.trailingAnchor],
+    [self.topAnchor constraintEqualToAnchor:self.topAnchor],
+    [self.bottomAnchor constraintEqualToAnchor:self.bottomAnchor],
+
+    [_stackView.leadingAnchor constraintEqualToAnchor:self.leadingAnchor],
+    [_stackView.trailingAnchor constraintEqualToAnchor:self.trailingAnchor],
+    [_stackView.widthAnchor constraintGreaterThanOrEqualToAnchor:self.widthAnchor],
+    [_stackView.topAnchor constraintEqualToAnchor:self.topAnchor],
+    [_stackView.bottomAnchor constraintEqualToAnchor:self.bottomAnchor],
+  ]];
```

<!-- identifier: clang-format -->
<!-- File: components/Tabs/src/TabBarView/MDCTabBarView.m -->

material-automation commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -85,32 +73,71 @@ - (void)setSelectedItem:(UITabBarItem *)selectedItem {
+  CGFloat requiredWidth = maxWidth * self.items.count;
+  BOOL canBeJustified = requiredWidth > availableWidth;
+  self.stackView.distribution = canBeJustified ? UIStackViewDistributionFillProportionally
+  : UIStackViewDistributionFillEqually;
```
clang-format suggested the following change:


```
-  : UIStackViewDistributionFillEqually;
+                                               : UIStackViewDistributionFillEqually;
```

<!-- identifier: clang-format -->
<!-- File: components/Tabs/src/TabBarView/MDCTabBarView.m -->

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

material-automation commented:
bazel detected changes to the following targets:

```
//components/Tabs:ObjcExamples
//components/Tabs:TabBarView
//components/Tabs:snapshot_test_lib
//components/Tabs:snapshot_tests
//components/Tabs:snapshot_tests.swift_runtime_linkopts
//components/Tabs:snapshot_tests_test_binary
//components/Tabs:snapshot_tests_test_bundle
//components/Tabs:unit_test_tabbarview_sources
//components/Tabs:unit_tests
//components/Tabs:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/Tabs:unit_tests_IPAD_PRO_12_9_IN_9_3.swift_runtime_linkopts
//components/Tabs:unit_tests_IPAD_PRO_12_9_IN_9_3_test_binary
//components/Tabs:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/Tabs:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/Tabs:unit_tests_IPHONE_7_PLUS_IN_10_3.swift_runtime_linkopts
//components/Tabs:unit_tests_IPHONE_7_PLUS_IN_10_3_test_binary
//components/Tabs:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/Tabs:unit_tests_IPHONE_X_IN_11_0
//components/Tabs:unit_tests_IPHONE_X_IN_11_0.swift_runtime_linkopts
//components/Tabs:unit_tests_IPHONE_X_IN_11_0_test_binary
//components/Tabs:unit_tests_IPHONE_X_IN_11_0_test_bundle
```

<!-- identifier: affected-targets -->

material-automation commented:
bazel detected changes to the following targets:

```
//components/Tabs:snapshot_tests
//components/Tabs:snapshot_tests_test_bundle
//components/Tabs:unit_tests
//components/Tabs:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/Tabs:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/Tabs:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/Tabs:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/Tabs:unit_tests_IPHONE_X_IN_11_0
//components/Tabs:unit_tests_IPHONE_X_IN_11_0_test_bundle
```

<!-- identifier: affected-targets -->

Commit:
Fix clang format

Commit:
Add snapshot tests

mikefan1991 commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -15,13 +15,15 @@
 #import "MDCTabBarView.h"
 #import "private/MDCTabBarViewItemView.h"
 
+#import "private/MDCTabBarViewItemView.h"
```
This is duplicate import.

romoore replied:
Removed.

mikefan1991 commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -30,10 +32,10 @@ @implementation MDCTabBarView
 
 - (instancetype)init {
-  self = [super init];
+  self = [super initWithFrame:CGRectZero];
```
This change is not necessary.

romoore replied:
Reverted.

mikefan1991 commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -93,28 +80,62 @@ - (void)setSelectedItem:(UITabBarItem *)selectedItem {
+    MDCTabBarViewItemView *itemView = [[MDCTabBarViewItemView alloc] init];
+    itemView.translatesAutoresizingMaskIntoConstraints = NO;
+    itemView.title = item.title;
+    itemView.image = item.image;
```
I have removed the title and the image in another PR. Please wait for that to be merged and then update this.

romoore replied:
Done.

mikefan1991 commented on components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m:
```diff
@@ -48,7 +48,7 @@ - (void)setUp {
   // Uncomment below to recreate all the goldens (or add the following line to the specific
   // test you wish to recreate the golden for).
-  //  self.recordMode = YES;
+  self.recordMode = YES;
```
This should stay commended.

romoore replied:
Done.

romoore commented:
Gonna take a stab at cleaning this up. Ping me if you plan to work on it as well.

romoore commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -30,10 +32,10 @@ @implementation MDCTabBarView
   if (self) {
     _items = @[];
-    _itemViews = @[];
+    [self setUpSubviews];
```
This looks far too complex to be called from `-init`.  Let's try to simplify and add the constraints later.

leonmz replied:
Fixed.

Commit:
Merge branch 'develop' into pr-7704

Commit:
Cleaning-up based on PR feedback.

* Remove duplicate import.
* Revert unnecessary use of `initWithFrame:`
* Turn off `recordMode` in snapshot test.

romoore commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -20,8 +20,8 @@
-/** The views representing the items of this tab bar. */
-@property(nonatomic, strong) NSArray<UIView *> *itemViews;
+/** The stack view that contains all tab items. */
+@property(nonatomic) UIStackView *stackView;
```
Nit: Please document retention.

leonmz replied:
Not sure if this change resolve this issue..

romoore replied:
I don't see a difference. Perhaps you need to push? I think it's also better to name this for its semantic use (rather than its type).  I think something like this would be better:

```objc
@property(nonnull, nonatomic, strong) UIStackView *containerView;
```

leonmz replied:
Fix applied!

romoore commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -93,28 +78,62 @@ - (void)setSelectedItem:(UITabBarItem *)selectedItem {
 }
 
-#pragma mark - UIView
+- (void)setUpSubviews {
```
Please rename this to `commonMDCTabBarViewInit`.

leonmz replied:
Fix applied.

romoore commented on components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m:
```diff
@@ -137,4 +137,63 @@ - (void)testItemsWithMixedTitlesAndImages {
   [self generateSnapshotAndVerifyForView:self.tabBarView];
 }
 
+- (void)testMultipleLineItemViewToFitInTheSingleScreen {
```
Please add a similar test that uses the bar's `intrinsicContentSize` to set the bounds so we can be sure it will return the correct size to fit the items.

leonmz replied:
Added.

romoore commented on components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m:
```diff
@@ -137,4 +137,63 @@ - (void)testItemsWithMixedTitlesAndImages {
+  [self generateSnapshotAndVerifyForView:self.tabBarView];
+}
+
+- (void)testTooManyItemsInScrollView {
```
Nit: Naming: Please remove "InScollView" from the test since the behavior is automatic and depends on bounds sizes, etc.  Consider the name `testItemsLargerThanBoundsChangesToScrollableLayout`?

leonmz replied:
Thanks for the advising!

romoore commented on components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m:
```diff
@@ -137,4 +137,63 @@ - (void)testItemsWithMixedTitlesAndImages {
+  [self generateSnapshotAndVerifyForView:self.tabBarView];
+}
+
+- (void)testSuperLongItemViewInScrollView {
```
A few things here: 
*  Please make the bounds larger than the single item's maximum width (I believe it's 360)
*  Please remove "InScrollView" from the name, since the layout depends on content + bounds and isn't guaranteed.  Perhaps `testVeryLongItemLimitsWidth` or something similar?

leonmz replied:
Added :)

Commit:
Fix the PR comments

material-automation commented on components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m:
```diff
@@ -137,4 +137,83 @@ - (void)testItemsWithMixedTitlesAndImages {
+  // Given
+  self.tabBarView.bounds = CGRectMake(0, 0, 720, kExpectedHeightTitlesAndIcons);
+  NSString *longString =
+    @"This is a super long tab bar string. And it should be longer than 360 and be multipl line.";
```
clang-format suggested the following change:


```
-    @"This is a super long tab bar string. And it should be longer than 360 and be multipl line.";
+      @"This is a super long tab bar string. And it should be longer than 360 and be multipl line.";
```

<!-- identifier: clang-format -->
<!-- File: components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m -->

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

Commit:
Clang format

Commit:
Clang format

Commit:
Merge branch 'layout' of https://github.com/leonmz/material-components-ios into layout

leonmz committed:
Merge branch 'develop' into layout

Commit:
Fix presubmit tests

Commit:
Fix presubmit tests

material-automation commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -99,32 +87,88 @@ - (void)setSelectedItem:(UITabBarItem *)selectedItem {
+  CGFloat requiredWidth = maxWidth * self.items.count;
+  BOOL canBeJustified = availableWidth > requiredWidth;
+  self.containerView.distribution = canBeJustified ? UIStackViewDistributionFillEqually
+                                               : UIStackViewDistributionFillProportionally;
```
clang-format suggested the following change:


```
-                                               : UIStackViewDistributionFillProportionally;
+                                                   : UIStackViewDistributionFillProportionally;
```

<!-- identifier: clang-format -->
<!-- File: components/Tabs/src/TabBarView/MDCTabBarView.m -->

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

Commit:
Merge branch 'layout' of https://github.com/leonmz/material-components-ios into layout

mikefan1991 commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -99,32 +87,88 @@ - (void)setSelectedItem:(UITabBarItem *)selectedItem {
   _selectedItem = selectedItem;
 }
 
-#pragma mark - UIView
```
#pragma mark - Private API should be added for the following functions.

leonmz replied:
Fixed.

Commit:
Add pragma mark

Commit:
Merge remote-tracking branch 'upstream/develop' into layout

leonmz commented:
I refactored the colorFortState function a bit. Deleted the items view and use the arranged view from containerView instead. Please take a look at it. Thanks!

romoore commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -20,8 +20,11 @@
+@property(nonnull, nonatomic, strong) UIStackView *containerView;
+
+/** The constraints managing this view. */
+@property(nonatomic, strong) NSArray *viewConstraints;
```
Please add a nullability annotation to this property. Please also use generics to document that it contains `NSLayoutConstraint *` types.

romoore replied:
Done. I was able to remove this property and replace it with a flag. #namingIsHard

romoore commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -142,30 +129,91 @@ - (UIColor *)titleColorForState:(UIControlState)state {
+    [_containerView.bottomAnchor constraintEqualToAnchor:self.bottomAnchor],
+  ]];
+
+  [self addConstraints:constraints];
```
Please don't call `addConstraints` since it will be deprecated. Instead, activate the constraints directly.

romoore replied:
Done

romoore commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -142,30 +129,91 @@ - (UIColor *)titleColorForState:(UIControlState)state {
-  CGSize itemSize = CGSizeMake(boundsSize.width / self.items.count, boundsSize.height);
+  CGFloat requiredWidth = maxWidth * self.items.count;
+  BOOL canBeJustified = availableWidth > requiredWidth;
+  self.containerView.distribution = canBeJustified ? UIStackViewDistributionFillEqually
```
It seems like `layoutSubviews` shouldn't be required if a constraint-based layout is used. Even then, I would expect that this code execute before `[super layoutSubviews]` since that is expected to use any constraints to layout the subviews.

I suspect a better place to configure `self.containerView.distribution` might be in `setItems` and in `setBounds`.

romoore replied:
I'll leave this here for now. `setBounds:` is called more times during the initial presentation of a view and not much less often during rotation. It's likely not worth the optimization.

romoore commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -142,30 +129,91 @@ - (UIColor *)titleColorForState:(UIControlState)state {
+- (void)setUpItemViews {
+  for (UIView *view in self.containerView.arrangedSubviews) {
+    [view removeFromSuperview];
+    [_containerView removeArrangedSubview:view];
```
This isn't necessary since `removeFromSuperview` automatically removes it from the stack view's `arrangedSubviews` property.

romoore replied:
Removed.

romoore commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -142,30 +129,91 @@ - (UIColor *)titleColorForState:(UIControlState)state {
 
-  if (self.items.count == 0) {
-    return;
+  CGFloat availableWidth = self.bounds.size.width;
```
Nit: Please use `CGRectGetWidth(self.bounds)` instead of accessing the bounds directly. Especially for UIScrollViews, the bounds origin may not be (0, 0).

romoore replied:
Done.

romoore commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -142,30 +129,91 @@ - (UIColor *)titleColorForState:(UIControlState)state {
   }
-  CGSize boundsSize = CGSizeMake(CGRectGetWidth(self.bounds), CGRectGetHeight(self.bounds));
-  CGSize itemSize = CGSizeMake(boundsSize.width / self.items.count, boundsSize.height);
+  CGFloat requiredWidth = maxWidth * self.items.count;
```
Looks like this can be replaced with a call to `self.intrinsicContentSize`?

romoore replied:
Ah, no, that returns the scrollable width, not the justified width.

romoore commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -142,30 +129,91 @@ - (UIColor *)titleColorForState:(UIControlState)state {
 - (CGSize)sizeThatFits:(CGSize)size {
-  return CGSizeMake(size.width, MAX(size.height, kMinHeight));
+  CGSize intrinsicSize = self.intrinsicContentSize;
+  return CGSizeMake(MIN(intrinsicSize.width, size.width), MAX(intrinsicSize.height, size.height));
```
I'm curious about this logic. If the current bounds is `CGRectZero` and `sizeToFit` is called, then this view would only resize its height. Is that expected?  Should it not attempt to provide a fit size for its contents as a minimum, but allow a larger width?

leonmz replied:
I am thinking about since it is a scrollable view thus this should be fits in any width.

leonmz replied:
CGZero is a extremely case, but if a small bounds is provided, the scrollable view should also fit itself in

romoore replied:
That's true, but the fit size would (ideally) be the size of its content. I think the best thing to conform to the documented behavior of `sizeThatFits:` is to return the "fit size".  In this case, it's probably the the size of the justified layout (rather than the intrinsic content size).  Does that make sense?

Whether it's kept like this or changed, I think it's worth adding a unit test that calls `sizeToFit:` with a bounds of `CGRectZero` and make sure it gets back a non-zero width and non-zero height.  Something similar for a starting bounds bigger than necessary?

leonmz replied:
Sure. Fixed applied. I also refactored the intrisicSize to justified width instead of scrollable width.

romoore commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -142,30 +129,91 @@ - (UIColor *)titleColorForState:(UIControlState)state {
+
+#pragma mark - Private Functions
+
+- (void)commonMDCTabBarViewInit {
```
Nit: Please move this up near the `-init` methods.

romoore replied:
Done. Moved into `-init`

romoore commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -142,30 +129,91 @@ - (UIColor *)titleColorForState:(UIControlState)state {
+  self.backgroundColor = UIColor.whiteColor;
+
+  [self setUpStackView];
+  [self setUpItemViews];
```
This call doesn't seem necessary.  I believe the `items` array is empty during initialization.

romoore replied:
Removed.

romoore commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -142,30 +129,91 @@ - (UIColor *)titleColorForState:(UIControlState)state {
+  [self addSubview:_containerView];
+}
+
+- (void)setUpItemViews {
```
I believe this logic can be put back in `setItems:` since it's only called from there.

leonmz replied:
I still feel it worth a function extract the logic..

romoore replied:
Could you help me understand your thinking? We cannot unit test this code and it's only ever called from within `setItems:`. Either testability or duplicate code would be good candidates for a method. In this case I'm trying to understand the value of an extra message send (plus complexity in figuring out what the method does).

leonmz replied:
Thanks  a lot for detailed explanation!!

romoore reviewed (changes requested):
I'm going to address most of my feedback and push up changes to the branch.

Commit:
Clean up and refactoring a bit.

Commit:
oops, flip the flag

Commit:
Restore position of loop

material-automation commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -142,30 +146,56 @@ - (UIColor *)titleColorForState:(UIControlState)state {
+  [_containerView.widthAnchor constraintGreaterThanOrEqualToAnchor:self.widthAnchor].active = YES;
+  [_containerView.topAnchor constraintEqualToAnchor:self.topAnchor].active = YES;
+  [_containerView.bottomAnchor constraintEqualToAnchor:self.bottomAnchor].active = YES;
+  self.containerViewConstraintsActive = YES;
```
clang-format suggested the following change:


```
-  
+
```

<!-- identifier: clang-format -->
<!-- File: components/Tabs/src/TabBarView/MDCTabBarView.m -->

romoore replied:
Done?

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

Commit:
Formatting

Commit:
Update example with better colors.

romoore commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -142,30 +147,56 @@ - (UIColor *)titleColorForState:(UIControlState)state {
+  [super updateConstraints];
 }
 
 - (CGSize)intrinsicContentSize {
```
Please add unit tests to cover verify the behavior/values of `intrinsicContentSize`.

leonmz replied:
Added.

romoore replied:
No, I don't think the TabBarView should be dependent on its contents.  I'd prefer the tests cover the more general behavior of `intrinsicContentSize`. For example, ensuring that it returns the justified/scrollable width consistently (depending on what you determine is more useful). That the height grows if one of its items has a very large image or long text.  It's the relationship/behavior that's important and not the exact values.

Let me add one or two as a commit and we can talk about them.  I think that will be faster and clearer than my trying to explain in a comment. ;-)

romoore commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -142,30 +147,56 @@ - (UIColor *)titleColorForState:(UIControlState)state {
+  return CGSizeMake(totalWidth, MAX(kMinHeight, maxHeight));
 }
 
 - (CGSize)sizeThatFits:(CGSize)size {
```
I also mentioned it in another thread, but please add unit tests for `sizeThatFits:`

leonmz replied:
Added.

romoore reviewed (changes requested):
Looks good, but needs unit tests for `intrinsicContentSize` and `sizeThatFits:` before it's ready for merge.

Commit:
Change IntrisicSize to Justified Width and add tests

leonmz commented:
Refactored the intrsicSize to justified size. Also added snapshot tests for case that the bounds is 1000 but only two items exist. The current behavior is that the item view has been stretched to 500 due to its low hugging priority.

Commit:
Merge branch 'develop' into pr-7704

Commit:
More unit tests for `intrinsicContentSize`. Renaming some tests.

Commit:
Refactoring and updating snapshot tests.

Commit:
Merge branch 'develop' into pr-7704

material-automation commented on components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m:
```diff
@@ -242,6 +248,120 @@ - (void)testChangingSelectedImageOfSelectedItemAfterAddingToBarDoesNothing {
+  // When
+  self.tabBarView.items = @[ item1, item2, item3, item4, item5, item6 ];
+  CGSize intrinsicContentSize = self.tabBarView.intrinsicContentSize;
+  self.tabBarView.bounds = CGRectMake(0, 0, intrinsicContentSize.width, intrinsicContentSize.height);
```
clang-format suggested the following change:


```
-  self.tabBarView.bounds = CGRectMake(0, 0, intrinsicContentSize.width, intrinsicContentSize.height);
+  self.tabBarView.bounds =
+      CGRectMake(0, 0, intrinsicContentSize.width, intrinsicContentSize.height);
```

<!-- identifier: clang-format -->
<!-- File: components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m -->

material-automation commented on components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m:
```diff
@@ -242,6 +248,120 @@ - (void)testChangingSelectedImageOfSelectedItemAfterAddingToBarDoesNothing {
+- (void)testVeryLongItemLimitsItemWidthToItemMaximumWhenBoundsTooNarrow {
+  // Given
+  NSString *longString =
+      @"This is a super long tab bar string. And it should be longer than 360 and wrap to multiple lines.";
```
clang-format suggested the following change:


```
-  NSString *longString =
-      @"This is a super long tab bar string. And it should be longer than 360 and wrap to multiple lines.";
+  NSString *longString = @"This is a super long tab bar string. And it should be longer than 360 "
+                         @"and wrap to multiple lines.";
```

<!-- identifier: clang-format -->
<!-- File: components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m -->

material-automation commented on components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m:
```diff
@@ -242,6 +248,120 @@ - (void)testChangingSelectedImageOfSelectedItemAfterAddingToBarDoesNothing {
+
+  // When
+  self.tabBarView.items = @[ item1, item2, item3 ];
+  self.tabBarView.bounds = CGRectMake(0, 0, kMaxItemWidth * self.tabBarView.items.count / 2, kExpectedHeightTitlesOrIconsOnly);
```
clang-format suggested the following change:


```
-  self.tabBarView.bounds = CGRectMake(0, 0, kMaxItemWidth * self.tabBarView.items.count / 2, kExpectedHeightTitlesOrIconsOnly);
+  self.tabBarView.bounds = CGRectMake(0, 0, kMaxItemWidth * self.tabBarView.items.count / 2,
+                                      kExpectedHeightTitlesOrIconsOnly);
```

<!-- identifier: clang-format -->
<!-- File: components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m -->

Commit:
Formatting

material-automation commented on components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m:
```diff
@@ -242,6 +248,120 @@ - (void)testChangingSelectedImageOfSelectedItemAfterAddingToBarDoesNothing {
+- (void)testVeryLongItemLimitsItemWidthToItemMaximumWhenBoundsAreIntrinsicContentSize {
+  // Given
+  NSString *longString =
+  @"This is a super long tab bar string. And it should be longer than 360 and wrap to multiple lines.";
```
clang-format suggested the following change:


```
-  NSString *longString =
-  @"This is a super long tab bar string. And it should be longer than 360 and wrap to multiple lines.";
+  NSString *longString = @"This is a super long tab bar string. And it should be longer than 360 "
+                         @"and wrap to multiple lines.";
```

<!-- identifier: clang-format -->
<!-- File: components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m -->

material-automation commented on components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m:
```diff
@@ -242,6 +248,120 @@ - (void)testChangingSelectedImageOfSelectedItemAfterAddingToBarDoesNothing {
+  self.tabBarView.items = @[ item1, item2, item3 ];
+  CGSize intrinsicContentSize = self.tabBarView.intrinsicContentSize;
+  self.tabBarView.bounds = CGRectMake(0, 0, intrinsicContentSize.width,
+                                      intrinsicContentSize.height);
```
clang-format suggested the following change:


```
-  self.tabBarView.bounds = CGRectMake(0, 0, intrinsicContentSize.width,
-                                      intrinsicContentSize.height);
+  self.tabBarView.bounds =
+      CGRectMake(0, 0, intrinsicContentSize.width, intrinsicContentSize.height);
```

<!-- identifier: clang-format -->
<!-- File: components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m -->

material-automation commented on components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m:
```diff
@@ -242,6 +248,120 @@ - (void)testChangingSelectedImageOfSelectedItemAfterAddingToBarDoesNothing {
+  self.tabBarView.items = @[ item1, item2 ];
+  CGSize intrinsicContentSize = self.tabBarView.intrinsicContentSize;
+  self.tabBarView.bounds = CGRectMake(0, 0, MIN(kMaxItemWidth * self.tabBarView.items.count,
+                                                intrinsicContentSize.width * 2), intrinsicContentSize.height);
```
clang-format suggested the following change:


```
-  self.tabBarView.bounds = CGRectMake(0, 0, MIN(kMaxItemWidth * self.tabBarView.items.count,
-                                                intrinsicContentSize.width * 2), intrinsicContentSize.height);
+  self.tabBarView.bounds = CGRectMake(
+      0, 0, MIN(kMaxItemWidth * self.tabBarView.items.count, intrinsicContentSize.width * 2),
+      intrinsicContentSize.height);
```

<!-- identifier: clang-format -->
<!-- File: components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m -->

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

Commit:
Don't use ivar outside init and accessors

Commit:
Fixing loss-of-precision errors

romoore commented:
@googlebot All authors are Googlers, but one commit has a misconfigured author email.
